### ISSUE TITLE
Increased loop time

### ISSUE THREAD

**Issue by mbraeunlein** [laurensvalk/segway#6]
Hi @laurensvalk,

first of all thank you for the code so far, the robot actually balances, even though it is not pretty smooth.
I also ran the balancer software you "coded" using the lego legacy software. (http://robotsquare.com/2014/07/01/tutorial-ev3-self-balancing-robot/) which makes the robot run much more stable while also being able to work with a loop time of 22 milliseconds. 

My computations for speed and steering require about 10ms, so it is currently not possible for me to use the python implementation.

The difference between the two implementations seems to be in the PID controller. I tried to reimplement the software but with no success yet.

Do you plan to release a ported version? Or do you have parameters that enable the robot to balance with a looptime of 20+ms?

Best,
Michael

**Comment by laurensvalk** [comment]
Hey @mbraeunlein, thanks for trying out the code!

The algorithm implemented in [Python](https://github.com/laurensvalk/segway/tree/master/ev3/ev3dev/python) is actually much simpler than the old [EV3-G program](http://robotsquare.com/2014/07/01/tutorial-ev3-self-balancing-robot/), which still contains some superfluous code. I still mean to update the EV3-G program to make it consistent with the simpler Python code, but since it's working well in its current state, this hasn't had a high priority.

What kind of computations are you doing to find speed and steering? Computations in the loop should generally be fine, but I have found that some hardware (like a bluetooth remote control) can seriously affect performance. Although the average loop time may seem alright, this led to some loops taking much longer than others because it was busy handling bluetooth.

Either way, the new Python program is meant to work at least as smooth as the EV3-G version ([Here's one video running the Python code](https://www.youtube.com/watch?v=_3Md3oUe008)) and going towards 20 ms should be no problem. Making it more smoothly generally amounts to adapting these values:

````
gainGyroAngle                  =   # For every radian (57 degrees) we lean forward,            apply this amount of duty cycle.
gainGyroRate                   =   # For every radian/s we fall forward,                       apply this amount of duty cycle.
gainMotorAngle                 =   # For every radian we are ahead of the reference,           apply this amount of duty cycle
gainMotorAngularSpeed          =   # For every radian/s drive faster than the reference value, apply this amount of duty cycle
gainMotorAngleErrorAccumulated =   # For every radian x s of accumulated motor angle,          apply this amount of duty cycle
````

This can be a bit tricky (but fun as you start to get the hang of it). Basically, increasing one number increases the sensitivity to that sensor value. So increasing gainGyroAngle makes it respond to a nonzero body angle with greater motor voltage. But increase it too much and it will try to over-compensate and become unstable. To get started, set the gains related to the motor angle to 0 and see how the robot responds to angular rate or angle, for example.

I'm not sure what the best way would be to make certain values work for everybody. Somehow, the EV3-G version seems to work for most people without adapting any values, so I'm not sure what is causing the inconsistencies with ev3dev/Python.

**Comment by mbraeunlein** [comment]
Hi @laurensvalk,

thank you for your fast response. 

The robot in the video looks really stable, that is about what i imagined to achieve. I use slightly taller wheels (62mm diameter) and mounted two color sensors at the bottom to be able to detect in which direction i cross a border. There is no bluetooth or fancy stuff involved, just the two sensor reads for which i also used the FastRead function and some if's.

As you mentioned [here](https://github.com/rhempel/ev3dev-lang-python/issues/130), your loop takes about 6ms. I measured the time right before the program sleeps for the rest of the loop  using
`...`
`print time.clock() - tLoopStart`
`while(time.clock() - tLoopStart <  loopTimeSec):`
`...`
and get values of 8-10ms with your code copy pasted and the robot is kind of shivering.

I was trying to get a feel for the parameters for the last weekend straight but did not get the robot to balance as stable as in your video.

With the EV3-G program i did not have any issues even if i did not adjust the wheel parameter. I also can not imagine what those differences come from so my guess was the PID controler. Why is it superfluous in your opinion?

Best,
Michael

**Comment by laurensvalk** [comment]
> I also can not imagine what those differences come from so my guess was the PID controler. Why is it superfluous in your opinion?

The old code was based on literature I found elsewhere, as referenced on the tutorial page. Since then, I studied control theory, and I found that the PID term was not necessary at all to achieve stabilization.

In the newer code, proportional control (P) is achieved with feedback on body angle and wheel angle. Damping (D) is now achieved by feedback to angular rate of body and the motors, while integral action (I) is achieved using the accumulated motor term angle.

Overal, there are now just 5 control terms, rather than 12 (four signals, each with three PID terms). Some of these 12 were partially overlapping or achieving an opposite effect. For example, derivative control on the body angle is the same as proportional control on the body angular rate. 

That doesn't do too much harm and computation time is not so much longer, but it makes tuning even less insightful than having just 5 parameters.
